### ISSUE TITLE
Failed to connect to an IPC Port: Access is denied

### ISSUE THREAD

**Issue by adal01** [EasyHook/EasyHook#126]
I have one more issue. Also, thanks for the help in the previous thread.

When I make the call  below I get the the following error.

"Failed to connect to an IPC Port: Access is denied.\r\n "

 RemoteHooking.Inject(
                    m_TargetPID,
                    "FileMonInject.dll",
                    "FileMonInject.dll",
                    m_ChannelName);

I didn't get this error when I ran the demo Filemon.exe, but now that I have replaced the filemon.exe with my own program PCTextCatpure.exe I now get the error. 

I would think it's a rights issue on my machine but the filemon.exe demo works without rights issues.

**Comment by spazzarama** [comment]
Both apps running in the same windows session? As same user etc..? Try debugging the target app and see if you can break point within the `EasyHook.IEntryPoint` of FileMonInject.dll (assuming you built it and have the source handy).

**Comment by adal01** [comment]
Yes both apps are running in the same windows session, and user session.

The FileMonInject.dll will inject and run in the target app if I use the demo app filemon.exe

The same file FileMonInject.dll won't run if I use my replacement app PCTextCapture.exe 

It seems the problem is somehow with my newly created app PCTextCatpure.exe. Since the FilemonInject.dll hasn't changed. Hmm......

This is the code, The line RemoteHooking.Inject(..... ) fails.
{
                // INSTALL THE TEXT CAPTURE HOOK
                Config.Register(
                       "A FileMon like demo application.",
                       "PCTextCapture.exe",
                       "FileMonInject.dll");

                RemoteHooking.IpcCreateServer<FileMonInterface>(ref m_ChannelName,   WellKnownObjectMode.SingleCall);

                RemoteHooking.Inject(
                    m_TargetPID,
                    "FileMonInject.dll",
                    "FileMonInject.dll",
                    m_ChannelName);

Great Project!

**Comment by adal01** [comment]
More information about the issue I'm having....

My newly created app "PCTextCapture.exe" that I created to replace the demo app "Filemon.exe" will NOT work with the sample "FilemonInject.dll" . Again the "FileMonInject.dll" hasn't change, and will work with the old example "Filemon.exe", but NOT the newly created "PCTextCapture.exe"

The error I get when trying to use the "PCTextCapture.exe" with the "FileMonInject.dll" is the following:

*** "Failed to connect to an IPC Port: Access is denied.\r\n " **

Everything works fine If I use the samples "FileMon.exe" with FileMonIInject.dll". I DO NOT get the error.

It seems it has something to do with creating a new app that then uses the "FileMonInject.dll."

Could it be some kind of signing issue? I noticed that the "EasyHook.dll" uses StrongName.snk, and both "FileMone.exe and "FileMonInject.dll" both use examples.snk.
 
Once again this is a great project!!!

**Comment by adal01** [comment]
Has anyone else had this problem? I get the same issue when I run my application at home and at work.
At this point I really don't thinks is a computer right issue. I thinks it a property embbeded in the FileMon.exe demo project. 

If I port all my code (From my PCTextCapture.exe) to the FileMon.exe demo project compile and run it. 
I don't not get the "Access is denied error"  It really does seem like it's a setting in the File.exe project.

Could it have something to do with signing the projects....?????? 

Out of ideas at this point....

**Comment by spazzarama** [comment]
the reason is because the IPC interface is defined in FileMon.exe rather than an external assembly, you will find that FileMonInject is loading FileMon.exe as a reference in the target and cannot use the correct IPC interface (since it knows nothing about it).

Either way you need to create your own inject assembly and either place the IPC class in there or reference whatever assembly you define it in.

Hope that helps.

**Comment by spazzarama** [comment]
To clarify, the class "FileMonInterface" from the FileMonInject project's perspective is in FileMon.exe not your new assembly.

**Comment by adal01** [comment]
The FileMonInterface is in my new assembly. I defined it in my new program.
Hmmm....

**Comment by spazzarama** [comment]
Yes, but did you recompile the FileMonInject assembly to use your one as well?

**Comment by adal01** [comment]
I added a reference from FileMonInject to my new assembly PCTextCapture.exe. I'm sure I did the recompile but I will double check. 

Thanks for the response

**Comment by spazzarama** [comment]
No I mean the FileMonInject assembly has a reference to FileMonInterface in FileMon.exe so you would need to change that assembly

**Comment by spazzarama** [comment]
Which is what you just said, sorry :)

**Comment by adal01** [comment]
just a little more information after playing around a little more last night.

After debugging in easyhooks, I found this is the call that fails when trying to register thee assemblies into the GAC.

IT FAILES HERE!!!!!!!!!
FILE DLLIMPORT.CS
public static void GacInstallAssemblies(
            String[] InAssemblyPaths,
            String InDescription,
            String InUniqueID)
        {
base {System.SystemException} = {System.UnauthorizedAccessException: Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED)) at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo) at System.Runtime.InteropServices.M...


I'm running Easyhooks version 2.7.5 when it fails....
All of these parameters seem to have the correct values before the call.
 String[] InAssemblyPaths,
            String InDescription,
            String InUniqueID

I run as admin, I have given all files and folders/files complete "everyone" privileges.

If I go and look at the GAC folder on my system at "C:\Windows\Microsoft.NET\assembly" I can see the assemblies have been added.

It even gets stranger! If I compile the same code under Easyhooks 2.5 everythings works OK! In other words, no problem with the GAC registration.

All of this happens before trying to use IPC so maybe that's where the real problem is in trying to register into the GAC.

**Comment by spazzarama** [comment]
I recommend not using the GAC registration at all, instead pass the file location to the assembly and the option DoNotRequireStrongNames into the inject call

**Comment by adal01** [comment]
Do I NOT make the call to the Config.Register call anymore?
Do I make the absolute path to the files in the RemoteHooking.Inject below, like shown.
 RemoteHooking.Inject(
                m_TargetPID,
                "c:\myfilelocation\FileMonInject.dll",
                "c:\myfilelocation\FileMonInject.dll",
                m_ChannelName);

Thanks in advance for your help.

**Comment by spazzarama** [comment]
Correct, that should do it for you

**Comment by spazzarama** [comment]
You can get the current assembly's path and use that too of course
